src/Config/ConfigMerger.php

```diff
@@ -57,13 +57,46 @@ public function isMergeRequired(array $config): bool
      * Merge two configs if merging is required, otherwise return $baseConfig without changes.
      *
      * @param array $baseConfig
-     * @param array $configToMerge
+     * @param array $customConfig
      * @return array
      */
-    public function mergeConfigs(array $baseConfig, array $configToMerge): array
+    public function merge(array $baseConfig, array $customConfig): array
     {
-        if ($this->isMergeRequired($configToMerge)) {
-            return array_replace_recursive($baseConfig, $this->clear($configToMerge));
+        if ($this->isMergeRequired($customConfig)) {
+            return array_replace_recursive($baseConfig, $this->clear($customConfig));
+        }
+
+        return $baseConfig;
+    }
+
+    /**
+     * Merges two configs according to rules.
+     *
+     * 1. If merge is required - merge configs
+     * 2. If specific key is required in $customConfig and present - merge configs
+     * 3. If no specific key is required and $customConfig not empty - merge configs
+     * 4. Otherwise return $baseConfig
+     *
+     * @param array $baseConfig
+     * @param array $customConfig
+     * @param string|null $key
+     * @return array
+     */
+    public function mergeIf(array $baseConfig, array $customConfig, string $key = null): array
```

Comment:
This method has no sense if needed to check the existence of more then one `key`. 
Checking of `keys` existence should be performed before calling `merge` method.

Reply:
This copy-paste is a sample of code duplication and bad design. The current method can be easily extended to include array of required keys if needed.

Reply:
For me, this method is a little bit strange. It named as `mergeIf` but can return `baseConfig` without merge,   `baseConfig` merged with `customConfig` or just `customConfig` without any merge. 
It is difficult to understand what will you get by calling this method.

Reply:
It has the same algorithm as another bunch of code duplicated places. It has the algorithm described in the description.

Reply:
I'm also really confused by the utility of this method, especially since the only place it's being used you aren't passing `$key`. In practice then, wouldn't this behave basically the same as `merge()`?

Reply:
(For the record, I've feel like the logic and usage of our ConfigMerger class has been a bit odd for a while now, so I'm all for making it clearer & more consistent, but I'm not sure this does that.)

Reply:
One more case, for example:
`$customConfig` is not empty and doesn't have option `_merge`, and provided `$key` not exists in 
`$customConfig` so result will be `$baseConfig` and this is not expected behavior for customers as they won't see any information why custom configuration wasn't applied.

Reply:
I propose to discuss this method offline ;) @shiftedreality can you book a room for tomorrow?

Reply:
I don't insist on having this method, will revert it back